core/src/storage2/collections/binary_heap/group.rs

```diff
@@ -0,0 +1,213 @@
+pub(crate) const COUNT: u32 = 2;
+
+/// Returns the index of the group in which the `n`-th element is stored.
+pub(crate) fn get_group_index(n: u32) -> u32 {
+    match n {
+        0 => 0,
+        _ => {
+            // The first group only ever contains the root element:
+            // `[Some(root), None]`. So when calculating indices we
+            // need to account for the items which have been left
+            // empty in the first group.
+            let padding = COUNT - 1;
+            (n + padding) / COUNT
+        }
+    }
+}
+
+/// Returns the in-group index of the `n`-th element.
+/// This refers to the index which the element has within a group.
+///
+/// For example, the element `3` is found at in-group index `0`
+/// (within the group at index `2`).
+pub(crate) fn get_ingroup_index(n: u32) -> Ingroup {
+    let group = get_group_index(n);
+    match (group, n) {
+        (0, 0) => Ingroup::Left,
+        (0, _) => panic!("first group contains only the root element"),
+        (_, _) => {
+            let ingroup_index = (n - 1) % COUNT;
+            match ingroup_index {
+                0 => Ingroup::Left,
+                1 => Ingroup::Right,
+                _ => {
+                    unreachable!(
+                        "COUNT is 2, following the modulo op index must be 0 or 1"
+                    )
+                }
+            }
+        }
+    }
+}
+
+impl<T> Group<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// Returns a shared reference to the element at `index`.
+    pub fn as_ref(&self, index: u32) -> Option<&T> {
+        let ingroup_index = get_ingroup_index(index);
+        match ingroup_index {
+            Ingroup::Left => self.0.as_ref(),
+            Ingroup::Right => self.1.as_ref(),
+        }
+    }
+
+    /// Returns an exclusive reference to the element at `index`.
+    pub fn get_mut(&mut self, index: u32) -> &mut Option<T> {
+        let ingroup_index = get_ingroup_index(index);
+        match ingroup_index {
+            Ingroup::Left => &mut self.0,
+            Ingroup::Right => &mut self.1,
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+const _: () = {
+    use crate::storage2::traits::StorageLayout;
+    use ink_metadata::layout2::{
+        CellLayout,
+        FieldLayout,
+        Layout,
+        LayoutKey,
+        StructLayout,
+    };
+
+    impl<T> StorageLayout for Group<T>
+    where
+        T: PackedLayout + Ord + TypeInfo + 'static,
+    {
+        fn layout(key_ptr: &mut KeyPtr) -> Layout {
+            StructLayout::new(vec![
+                FieldLayout::new(
+                    None,
+                    CellLayout::new::<i32>(LayoutKey::from(key_ptr.advance_by(1))),
+                ),
+                FieldLayout::new(
+                    None,
+                    CellLayout::new::<i64>(LayoutKey::from(key_ptr.advance_by(1))),
+                ),
+            ])
+            .into()
+        }
+    }
+};
+
+impl<T> SpreadLayout for Group<T>
```

Comment:
The whole point behind `Group` is that it is `AlwaysPacked` and thus always stores its fields as packed representation. The same as other data structures do with their internals. So their `SpreadLayout` impls mostly forward to the `PackedLayout` impl and their `FOOTPRINT` is always 1.

Reply:
Please take a deep look at examples from other parts of `ink_storage`, e.g. the `Stash`.

Reply:
Inspiration from `Stash` `Header` type, that is also `AlwaysPacked`:
```rust
impl SpreadLayout for Header {
    const FOOTPRINT: u64 = 1;
    const REQUIRES_DEEP_CLEAN_UP: bool = false;

    fn pull_spread(ptr: &mut KeyPtr) -> Self {
        forward_pull_packed::<Self>(ptr)
    }

    fn push_spread(&self, ptr: &mut KeyPtr) {
        forward_push_packed::<Self>(self, ptr)
    }

    fn clear_spread(&self, ptr: &mut KeyPtr) {
        forward_clear_packed::<Self>(self, ptr)
    }
}
```